src/pset/mod.rs

```diff
@@ -0,0 +1,237 @@
+        match self.global.tx_data {
+            // v0 PSETs are easy because they have an explicit transaction
+            GlobalTxData::V0 { ref unsigned_tx } => unsigned_tx.txid(),
+            GlobalTxData::V2 { .. } => unimplemented!(),
+        }
+    }
+
+    /// Extract the Transaction from a PartiallySignedTransaction by filling in
+    /// the available signature information in place.
+    pub fn extract_tx(self) -> Transaction {
+        let mut tx = match self.global.tx_data {
+            // v0 PSETs are easy because they have an explicit transaction
+            GlobalTxData::V0 { unsigned_tx } => unsigned_tx,
+            GlobalTxData::V2 { .. } => unimplemented!(),
+        };
+
+        for (vin, psetin) in tx.input.iter_mut().zip(self.inputs.into_iter()) {
+            vin.script_sig = psetin.final_script_sig.unwrap_or_else(Script::new);
+            vin.witness.script_witness = psetin.final_script_witness.unwrap_or_else(Vec::new);
+        }
+
+        tx
+    }
+
+    /// Attempt to merge with another `PartiallySignedTransaction`.
+    pub fn merge(&mut self, other: Self) -> Result<(), self::Error> {
+        if self.unique_id() != other.unique_id() {
+            return Err(Error::UniqueIdMismatch {
+                expected: self.unique_id(),
+                actual: other.unique_id(),
+            });
+        }
+
+        self.global.merge(other.global)?;
+
+        for (self_input, other_input) in self.inputs.iter_mut().zip(other.inputs.into_iter()) {
+            self_input.merge(other_input)?;
+        }
+
+        for (self_output, other_output) in self.outputs.iter_mut().zip(other.outputs.into_iter()) {
+            self_output.merge(other_output)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Encodable for PartiallySignedTransaction {
+    fn consensus_encode<S: io::Write>(
+        &self,
+        mut s: S,
+    ) -> Result<usize, encode::Error> {
+        let mut len = 0;
+        len += b"pset".consensus_encode(&mut s)?;
+
+        len += 0xff_u8.consensus_encode(&mut s)?;
+
+        len += self.global.consensus_encode(&mut s)?;
+
+        for i in &self.inputs {
+            len += i.consensus_encode(&mut s)?;
+        }
+
+        for i in &self.outputs {
+            len += i.consensus_encode(&mut s)?;
+        }
+
+        Ok(len)
+    }
+}
+
+impl Decodable for PartiallySignedTransaction {
+    fn consensus_decode<D: io::Read>(mut d: D) -> Result<Self, encode::Error> {
+        let magic: [u8; 4] = Decodable::consensus_decode(&mut d)?;
+
+        if *b"pset" != magic {
+            return Err(Error::InvalidMagic.into());
+        }
+
+        if 0xff_u8 != u8::consensus_decode(&mut d)? {
+            return Err(Error::InvalidSeparator.into());
+        }
+
+        let global: Global = Decodable::consensus_decode(&mut d)?;
+
+        let inputs: Vec<Input> = {
+            let inputs_len = global.n_inputs();
+
+            let mut inputs: Vec<Input> = Vec::with_capacity(inputs_len);
```

Comment:
You need to length-check `inputs_len` here (or perhaps above when decoding `global.n_inputs`?). Ditto for the code below regarding outputs.

Reply:
There is a sanity_check function which is called at the end that does all the checks.

Reply:
So, this saves allocation in case there is a mismatch? Does it matter if we are returning an Error anyways? In any case, I will change it.

Reply:
I don't follow what you are suggesting? How is possible to check this without the allocation.

Reply:
By defining a maximum value and then directly comparing it :)